Simplify h2 enable disable

shinrich opened the pull request:
This is work to address the issues that @zwoop raises in issue #5718 

As suggested, the first commit separates and enable/disable bitmask from the the protocol control continuations.  The protocol control continuations (Accept objects) are always created for all protocols regardless of whether they are enabled via the ports or not.

Then separate bit fields are constructed that track which protocols are enabled by the ports or the SNI.yaml file for the specific connection.

The second commit here implements the h2 action in the SNI.yaml file to replay the disable_h2 action.  Still leaving the disable_h2 action but marking it deprecated in preference to the more generation h2 action.  Added a h2enable.test.py in addition to the h2disable.test.py.

Still to do:  
* Rethink the InkAPI calls that do the same thing.  We used these calls originally via the tls_proto plugin to do the H2 disable before the core feature was available.
* Probably creating the npn more often than we need to.  Should be able to pull a value out of the accept object and only create a unique npn string in the h2 action occurs.

Also need to figure out the accept thread issue.

shinrich commented:
Pushed another commit to add tests to exercise the accept thread case.  Looks like the restructuring addressed the issue described in issue #5720

shinrich commented:
Proposed InkAPI change sent via the dev mailing list.  Copy here.

The TSProtoSet functions (https://docs.trafficserver.apache.org/en/latest/developer-guide/api/functions/TSProtoSet.en.html) were created for tls_proto, our early plugin to disable HTTP/2 on a per connection basis.  Since then, this HTTP/2 disable logic has been absorbed into the core via the sni.yaml, and with PR #5769 the underlying support for protocol enabling and disabling is simplified.  There, I propose replacing the existing TSProtoSet InkAPI calls with a slimmed down interface.

In the current InkAPI, to disable HTTP/2 for a SSL connection that has HTTP/2 enabled by default, the plugin would make the following calls.
```
TSAcceptor acceptor = TSAcceptorGet(sslp)
TSNextProtocolSet set = TSGetcloneProtoSet(acceptor)
set = TSUnregisterProtocol(set, "http/2")
TSRegisterProtocol(sslp, set)
```
Instead I propose the following pair of functions
```
TSReturnCode
TSVConnProtocolEnable(TSVConn connp, const char *protocol_name)

TSReturnCode
TSVConnProtocolDisable(TSVConn connp, const char *protocol_name)
```
If TSVConn is a SSLNetVC, then the alpn/npn string will be adjusted to add/remove the specified protocol.  If the protocol name is not registered, the function will return TSError.

If the TSVConn is a UnixNetVC, the functions will do nothing and return success.

In the future for HTTP/3, the functions could add/remove the protocols for however protocols are advertised for HTTP/3 or QUIC.

maskit commented on iocore/net/P_SNIActionPerformer.h:
```diff
@@ -56,24 +45,26 @@ class ActionItem
-      }
+    auto ssl_vc = dynamic_cast<SSLNetVConnection *>(cont);
+    if (control == YamlSNIConfig::Control::DISABLE) {
+      ssl_vc->disableProtocol(TS_ALPN_PROTOCOL_INDEX_HTTP_2_0);
```
ssl_vc can be nullptr

shinrich replied:
Not entirely sure that it can be.  I don't think these actions will be invoked outside of the SSL callbacks.  But better to be safe than sorry.  At least it will future proof the logic.

shinrich commented:
@maskit  Are there some instructions for building the QUIC feature?  A configure option or something?

shinrich commented:
Pushed another commit that integrates with the QUIC build.  Since the API is also updated, I removed the WIP label.  I can squash or separate the commits into separate PRs and people prefer.

maskit commented on iocore/net/QUICNetVConnection.cc:
```diff
@@ -503,6 +503,11 @@ QUICNetVConnection::free(EThread *t)
 
     super::clear();
   */
+  if (npn) {
```
Please use `this->`.

maskit commented on iocore/net/QUICNetVConnection.cc:
```diff
@@ -1018,6 +1023,27 @@ void
 QUICNetVConnection::registerNextProtocolSet(SSLNextProtocolSet *s)
 {
   this->_next_protocol_set = s;
+  this->_next_protocol_set->create_npn_advertisement(this->_protoenabled, &npn, &npnsz);
```
Please use `this->`.

maskit commented on iocore/net/P_QUICNetVConnection.h:
```diff
@@ -236,6 +240,10 @@ class QUICNetVConnection : public UnixNetVConnection, public QUICConnection, pub
   SSLNextProtocolSet *_next_protocol_set = nullptr;
+  SessionProtocolSet _protoenabled;
+  // Local copies of the npn strings
+  unsigned char *npn = nullptr;
```
Please put an underscore to private members.

maskit commented on iocore/net/quic/QUICConnection.h:
```diff
@@ -42,11 +43,14 @@ class QUICConnectionInfoProvider
+  virtual NetVConnectionContext_t direction() const               = 0;
+  virtual SSLNextProtocolSet *next_protocol_set() const           = 0;
+  virtual const SessionProtocolSet &get_enabled_protocols() const = 0;
+  virtual int select_next_protocol(SSL *ssl, const unsigned char **out, unsigned char *outlen, const unsigned char *in,
```
This should be `const`.

maskit commented on iocore/net/P_QUICNetVConnection.h:
```diff
@@ -171,6 +171,9 @@ class QUICNetVConnection : public UnixNetVConnection, public QUICConnection, pub
   // QUICNetVConnection
   void registerNextProtocolSet(SSLNextProtocolSet *s);
 
+  int select_next_protocol(SSL *ssl, const unsigned char **out, unsigned char *outlen, const unsigned char *in,
```
I want to put this line here, but for now, please move it under `// QUICConnection` since it's defined in QUICConnectionInfoProvider.

maskit commented on src/traffic_server/traffic_server.cc:
```diff
@@ -1969,7 +1969,9 @@ main(int /* argc ATS_UNUSED */, const char **argv)
         start_HttpProxyServer(); // PORTS_READY_HOOK called from in here
       }
     }
+#ifdef OLD
```
Should we keep this?

shinrich replied:
Nope, missed that one.  Will take it out.

maskit commented on src/traffic_server/InkAPI.cc:
```diff
@@ -7078,7 +7078,7 @@ TSNetAccept(TSCont contp, int port, int domain, int accept_threads)
 /* From proxy/http/HttpProxyServerMain.c: */
 extern bool ssl_register_protocol(const char *, Continuation *);
-extern bool ssl_unregister_protocol(const char *, Continuation *);
+// extern bool ssl_unregister_protocol(const char *, Continuation *);
```
Remove?

shinrich replied:
Yes

maskit commented:
Also, it seems like protocol negotiation is not working on QUIC connection (tested with traffic_quic).

maskit commented:
It seems like `QUICNetVC::_protoenabled` is not set at all.

Maybe it's the time to introduce `ALPNSupport` class that have `setEnabledProtocols`, `enableProtocol`  and `disableProtocol` ? Then we would be able to implement `TSVConnProtocolEnable` in a more generic way.

```cpp
class SSLNetVConnection : public UnixNetVConnection, public ALPNSupport ...
class QUICNetVConnection : public UnixNetVConnection, public ALPNSupport ...
```
```cpp
  auto alpn_support = dynamic_cast<ALPNSupport *>(net_vc);               
  if (alpn_support) {                                                                  
    alpn_support->enableProtocol(protocol_idx);                                        
    retval = TS_SUCCESS;                                                         
  }  
```

maskit commented on iocore/net/P_SSLNetVConnection.h:
```diff
@@ -161,6 +162,15 @@ class SSLNetVConnection : public UnixNetVConnection
+  void enableProtocol(int idx);
+
+  void
+  setEnabledProtocols(const SessionProtocolSet &protos)
```
It might make more sense if it was `setSupportedProtocols` and the variable name was `proto_supported`. It was hard to understand `protoenabled` contains protocols disabled and those are just marked as disabled.

shinrich replied:
The npnset contains all the protocols.  The protocolenabled is just a bit array that marks the protocols that should be enabled.  So I think that enabled is the more appropriate name than supported.  Could rename the npnset variable to be some variant of supported protocols.

shinrich commented:
Pushed up another commit to address the underscore issues and forgotten commented out bits.  Also fixed the unit test.  My compiler is evidently more lax.  Also fixed the uninitialized _protoenabled problem.  Was missing the calls to setEnabledProtocols in QUICNextProtocolAccept's main Accept event.  I don't know how to exercise quic, so please let me know if it works now.

I agree that refactoring to pull out comment logic for the npn/apln support and protocol management makes a lot of sense.  A lot of cut-n-paste here.  Let me take a pass on that.

randall commented on iocore/net/YamlSNIConfig.cc:
```diff
@@ -92,6 +92,7 @@ TsEnumDescriptor LEVEL_DESCRIPTOR         = {{{"NONE", 0}, {"MODERATE", 1}, {"ST
 TsEnumDescriptor POLICY_DESCRIPTOR        = {{{"DISABLED", 0}, {"PERMISSIVE", 1}, {"ENFORCED", 2}}};
 TsEnumDescriptor PROPERTIES_DESCRIPTOR    = {{{"NONE", 0}, {"SIGNATURE", 0x1}, {"NAME", 0x2}, {"ALL", 0x3}}};
 TsEnumDescriptor TLS_PROTOCOLS_DESCRIPTOR = {{{"TLSv1", 0}, {"TLSv1_1", 1}, {"TLSv1_2", 2}, {"TLSv1_3", 3}}};
+TsEnumDescriptor CONTROL_DESCRIPTOR       = {{{"NONE", 0}, {"ENABLE", 1}, {"DISABLE", 2}}};
```
What does NONE  do?

Why can't this be a boolean? Like @zwoop had in https://github.com/apache/trafficserver/pull/5719 ?

shinrich replied:
Internally we do need an unset value so we can determine whether the h2 field in the item class was set or not.  But you're right we don't need to advertise that to the end user, so I updated the CONTROL_DESCRIPTOR and the docs to remove the NONE option.

Now whether we want http2: on/off or h2: ENABLE/DISABLE or http/2: true/false, I don't really care.

shinrich commented:
Don't know why my branch is picking up changes to README.md and bootstrap.py under tests.  Not my intention.  Trying to figure out/undo that now.

shinrich commented:
I think this PR should be in good shape now.  Adjusted the sni.yaml interface to use http2:on/off.  I think the QUIC integration is working and moved some common code into ALPNSupport class.  Got rid of the irrelevant file changes.

maskit commented on src/traffic_server/InkAPI.cc:
```diff
@@ -9333,25 +9331,32 @@ TSSslTicketKeyUpdate(char *ticketData, int ticketDataLen)
+  TSReturnCode retval = TS_ERROR;
+  int protocol_idx    = globalSessionProtocolNameRegistry.toIndexConst(std::string_view{protocol_name});
+  auto net_vc         = reinterpret_cast<UnixNetVConnection *>(connp);
+  auto ssl_vc         = dynamic_cast<SSLNetVConnection *>(net_vc);
```
Shouldn't we cast `net_vc` to `ALPNSupport *`?

maskit commented:
`make check` passes now, H3 works as well.
You could test H3 without installing anything. You should have "traffic_quic" in your bin directory. The detail is on [the wiki page](https://cwiki.apache.org/confluence/display/TS/QUIC).

Why is `protoenabled` set twice, once in `setEnabledProtocols` and once in `registerNextProtocolSet`? Do we still need `setEnabledProtocols`?

maskit commented on iocore/net/P_SNIActionPerformer.h:
```diff
@@ -45,24 +42,28 @@ class ActionItem
+    if (ssl_vc) {
+      if (!enable_h2) {
+        ssl_vc->disableProtocol(TS_ALPN_PROTOCOL_INDEX_HTTP_2_0);
+      } else if (enable_h2) {
```
`enable_h2` is bool. Simple `else` should be enough.

maskit commented on iocore/net/ALPNSupport.cc:
```diff
@@ -0,0 +1,82 @@
+#include "records/I_RecHttp.h"
+
+void
+ALPNSupport::clear()
```
Hmm, I can live with this because NetVC already has `clear`, but I prefer to use its destructor if it's possible.

shinrich replied:
Will the destructor get called for a free list allocated object?

maskit replied:
Not automatically, but we should call it explicitly. Otherwise destructors for member objects won't be called neither. Of course we should call constructors too for free list allocated objects too, many places don't do it though.

shinrich commented:
Pushed a new version to address @maskit's comments.

shinrich commented:
[approve ci autest]
